### DISCUSSION TITLE
Function invocations as predicates

### DISCUSSION THREAD

**Discussion by gisborne** [cozodb/cozo#59] [Ideas]
There is a whole new world available if we can have “triggers” on predicates.

These could be defined by a URL — when a predicate is asserted, its arguments are sent to the URL.

These URLs might be “local” — in the limit, they might be functions made available in the server itself, through eg an embedded V8, or might be local processes connected through a socket. Note that sending a request to a HTTP endpoint and invoking a function have the same semantics: we send a set of key-value pairs.

This is particularly useful for [distributed systems](https://frest.substack.com/p/distributed-relational).

**Comment by gisborne** [comment]
Addendum: Just noticed where at the bottom of one page it says you can use callbacks from embedding languages.

This is adequate for my needs. Suggest it would be nice to have the server version support sending newly-ground rows to a URL.

**Reply by zh217** [reply]
Good idea, such callbacks can eliminate the need for a long-running service whose sole purpose is to relay the updates.
